### DISCUSSION TITLE
Adding post build step

### DISCUSSION THREAD

**Discussion by ukoda** [arendst/Tasmota#19837] [Q&A] [answered]
I have Tasmota building custom binaries to suit my environment and to reduce deployment times etc.  After building I need to copy (scp) the resulting binary files from my development PC to other systems such as my ESP programmer and servers etc.

I am currently doing that manually but would like to add a step the post build process.  Is there an existing hook into the build system, such as the user ini files, where I can do this, or do I need to dig into the build system and modify there?

**Comment by barbudor** [comment]
I'm doing similar process using tasmota-docker
I have a bash script ran nighly by cron on my home server which:
- pull the latest dev branch from arendst/tasmota
- pull the latest user_config_override.h and platformio_tasmota_cenv.ini from my private repo
- run the compile of my binaries
- move the binaries to my NAS

So I don't have to tweak platformio stuff

Even without using docker you can also have just a script that runs platformio builds and then move the binaries
Instead of integrating into platformio, do the opposite and integrate platformio in your script

**Reply by Noschvie** [reply]
Hi @barbudor : why do you have a "private" platformio_tasmota_cenv.ini and how do you handle changes from the repro? Thanks

**Reply by barbudor** [reply]
Because it defines my own custom builds, some being specific to a given site. So the binary has a different name. 
The 'c' in '_cenv' means custom and this file is not part of Tasmota repo. Same as user_config_override.h, there is a sample in the repo to help you create your own

**Comment by ukoda** [comment]
@barbudor Thanks for that suggestion.  I'm not using docker and I am still actively adding customisation to Tasmota for the devices I work with, so would prefer to have the bin files pushed for each device build run in the IDE.  Regardless your suggestion is a valid solution that I will call Plan B.

**Comment by Jason2866** [comment]
@ukoda  The Platformio OTA uploader script should do what you want. https://github.com/arendst/Tasmota/blob/development/pio-tools/http-uploader.py
or
https://github.com/arendst/Tasmota/blob/development/pio-tools/sftp-uploader.py

**Reply by ukoda** [reply, marked as answer]
@Jason2866 Thanks for that.  Those weren't quite the right files and initially I couldn't see how to fit that in with only modifying user files rather than files under version control.  However they did help me understand better how the build system works and where to look for what I needed.  From that I have come up with an incomplete solution, but the remaining works is just some Python code I need to write.

For others who find this thread the answer for me is to add an 'extra_scripts' declaration to the target in my platformio_tasmota_cenv.ini.  In my case I am building a modified version of tasmota.bin called ematasmota.bin and use this section in platformio_tasmota_cenv.ini:
```
[env:ematasmota]
build_flags                 = ${env.build_flags} -DFIRMWARE_EMA
extra_scripts               = ${esp_defaults.extra_scripts}
                              post:ema_upload.py

```
The first line of the extra script takes the final build steps and the third line adds a call to a new python script to it to do what I need after them.  My python script  looks like:
```
Import("env")

import os
import shutil
import tasmotapiolib

print("EMA upload of build binary")

def post_program_action(source, target, env):
    bin_file = tasmotapiolib.get_final_bin_path(env)
    print(f'EMA upload: bin_file={bin_file}')

env.AddPostAction("$BUILD_DIR/${PROGNAME}.bin", post_program_action)
```
This python code is incomplete, it currently just shows the file I will be transfering to where I need, but I still need to write that bit.

My thanks to the people who took time to reply to this thread :-)
